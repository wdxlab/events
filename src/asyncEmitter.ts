import { BaseEmitter, EmitterOptions, FilterFn, Handler, MapFn } from './baseEmitter';
import { AggregateError } from './aggregateError';

export type AsyncHandler<TArg> = Handler<TArg, Promise<void>>;

export class AsyncEmitter<TInputArg, TOutputArg = TInputArg> extends BaseEmitter<
  TInputArg,
  TOutputArg,
  Promise<unknown[]>
> {
  constructor(options?: EmitterOptions<TInputArg, TOutputArg, Promise<unknown[]>>) {
    super(options);

    this.sourceEmitter?.on(async (arg) => {
      const result = await this.emitInt(arg);
      if (result.length) {
        throw new AggregateError(result);
      }
    });
  }

  async emit(arg: TInputArg): Promise<unknown[]> {
    if (this.isReadOnly) {
      throw new Error('Emitter is readonly');
    }

    return this.emitInt(arg);
  }

  protected async emitInt(arg: TInputArg): Promise<unknown[]> {
    if (this.filterFn && !this.filterFn(arg)) {
      return [];
    }

    const errors: unknown[] = [];

    for (const fn of this.subscribers) {
      try {
        if (this.mapFn) {
          await fn(this.mapFn(arg));
        } else {
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          await fn(arg as any);
        }
      } catch (e) {
        errors.push(e);
        if (this.isBailMode) {
          break;
        }
      }
    }

    return errors;
  }

  filter(filter: FilterFn<TOutputArg>): AsyncEmitter<TOutputArg, TOutputArg> {
    return new AsyncEmitter<TOutputArg, TOutputArg>({
      source: this,
      filter,
      bail: this.isBailMode,
    });
  }

  map<TNewArg>(mapFn: MapFn<TOutputArg, TNewArg>): AsyncEmitter<TOutputArg, TNewArg> {
    return new AsyncEmitter<TOutputArg, TNewArg>({
      source: this,
      map: mapFn,
      bail: this.isBailMode,
    });
  }
}
