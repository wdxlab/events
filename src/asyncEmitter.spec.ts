import { AsyncEmitter } from './asyncEmitter';

it('should work', async () => {
  const emitter = new AsyncEmitter<{ name: string }>();
  const handler = jest.fn(() => new Promise((r) => setTimeout(r, 500)));
  const handler2 = jest.fn(() => new Promise((r) => setTimeout(r, 200)));

  emitter.on(handler);
  emitter.on(handler2);

  await emitter.emit({ name: 'foo' });
  expect(handler.mock.calls).toMatchSnapshot();
  expect(handler.mock.results).toMatchSnapshot();
  expect(handler2.mock.calls).toMatchSnapshot();
  expect(handler2.mock.results).toMatchSnapshot();

  await emitter.emit({ name: 'bar' });
  expect(handler.mock.calls).toMatchSnapshot();
  expect(handler.mock.results).toMatchSnapshot();
  expect(handler2.mock.calls).toMatchSnapshot();
  expect(handler2.mock.results).toMatchSnapshot();
});

describe('error', () => {
  test('non-bail', async () => {
    const emitterArg = { foo: 'hello' };
    const emitter = new AsyncEmitter<{ foo: string }>();
    const handler1 = jest.fn();
    const handler2 = jest.fn(() => {
      throw new Error('Foo');
    });
    const handler3 = jest.fn();

    emitter.on(handler1);
    emitter.on(handler2);
    emitter.on(handler3);

    const errors = await emitter.emit(emitterArg);

    expect(errors).toHaveLength(1);
    expect(errors[0]).toBeInstanceOf(Error);
    expect(handler1).toBeCalledTimes(1);
    expect(handler2).toBeCalledTimes(1);
    expect(handler3).toBeCalledTimes(1);
  });

  test('bail', async () => {
    const emitterArg = { foo: 'hello' };
    const emitter = new AsyncEmitter<{ foo: string }>({ bail: true });
    const handler1 = jest.fn();
    const handler2 = jest.fn(() => {
      throw new Error('Foo');
    });
    const handler3 = jest.fn();

    emitter.on(handler1);
    emitter.on(handler2);
    emitter.on(handler3);

    const errors = await emitter.emit(emitterArg);

    expect(errors).toHaveLength(1);
    expect(errors[0]).toBeInstanceOf(Error);
    expect(handler1).toBeCalledTimes(1);
    expect(handler2).toBeCalledTimes(1);
    expect(handler3).toBeCalledTimes(0);
  });
});

test('readonly mode', async () => {
  const source = new AsyncEmitter<number>();
  const target = new AsyncEmitter<number>({ source });
  const handler1 = jest.fn();
  const handler2 = jest.fn(() => {
    throw new Error('Foo');
  });
  const handler3 = jest.fn();

  target.on(handler1);
  target.on(handler2);
  target.on(handler3);

  await expect(target.emit(123)).rejects.toThrow('Emitter is readonly');

  const errors = await source.emit(456);

  expect(errors).toHaveLength(1);
  expect(errors[0]).toBeInstanceOf(Error);
  expect(handler1).toBeCalledTimes(1);
  expect(handler2).toBeCalledTimes(1);
  expect(handler3).toBeCalledTimes(1);
});

it('filterable', async () => {
  const emitter = new AsyncEmitter<{ name: string }>();
  const aaaFilter = emitter.filter((arg) => arg.name === 'aaa');
  const bbbFilter = emitter.filter((arg) => arg.name === 'bbb');
  const fooFilter = emitter.filter((arg) => arg.name === 'foo');
  const handler1 = jest.fn();
  const handler2 = jest.fn();

  emitter.on(handler1);
  aaaFilter.on(handler2);
  bbbFilter.on(handler2);
  fooFilter.on(handler2);

  await emitter.emit({ name: 'ccc' });
  expect(handler1.mock.calls).toMatchSnapshot();
  expect(handler2.mock.calls).toMatchSnapshot();
  await emitter.emit({ name: 'ddd' });
  expect(handler1.mock.calls).toMatchSnapshot();
  expect(handler2.mock.calls).toMatchSnapshot();
  await emitter.emit({ name: 'foo' });
  expect(handler1.mock.calls).toMatchSnapshot();
  expect(handler2.mock.calls).toMatchSnapshot();
});

test('map', async () => {
  type IArg = { event: { name: string }; source: { id: number } };
  const emitter = new AsyncEmitter<IArg>();
  const handler1 = jest.fn();

  emitter
    .filter((arg) => arg.source.id === 123)
    .map((arg) => arg.event)
    .filter((arg) => arg.name === 'foo')
    .map((arg) => arg.name)
    .on(handler1);

  await emitter.emit({ source: { id: 123 }, event: { name: 'bar' } });
  await emitter.emit({ source: { id: 456 }, event: { name: 'foo' } });
  await emitter.emit({ source: { id: 123 }, event: { name: 'foo' } });

  expect(handler1.mock.calls).toMatchSnapshot();
});
