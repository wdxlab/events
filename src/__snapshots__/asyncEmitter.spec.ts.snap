// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filterable 1`] = `
[
  [
    {
      "name": "ccc",
    },
  ],
]
`;

exports[`filterable 2`] = `[]`;

exports[`filterable 3`] = `
[
  [
    {
      "name": "ccc",
    },
  ],
  [
    {
      "name": "ddd",
    },
  ],
]
`;

exports[`filterable 4`] = `[]`;

exports[`filterable 5`] = `
[
  [
    {
      "name": "ccc",
    },
  ],
  [
    {
      "name": "ddd",
    },
  ],
  [
    {
      "name": "foo",
    },
  ],
]
`;

exports[`filterable 6`] = `
[
  [
    {
      "name": "foo",
    },
  ],
]
`;

exports[`map 1`] = `
[
  [
    "foo",
  ],
]
`;

exports[`should work 1`] = `
[
  [
    {
      "name": "foo",
    },
  ],
]
`;

exports[`should work 2`] = `
[
  {
    "type": "return",
    "value": Promise {},
  },
]
`;

exports[`should work 3`] = `
[
  [
    {
      "name": "foo",
    },
  ],
]
`;

exports[`should work 4`] = `
[
  {
    "type": "return",
    "value": Promise {},
  },
]
`;

exports[`should work 5`] = `
[
  [
    {
      "name": "foo",
    },
  ],
  [
    {
      "name": "bar",
    },
  ],
]
`;

exports[`should work 6`] = `
[
  {
    "type": "return",
    "value": Promise {},
  },
  {
    "type": "return",
    "value": Promise {},
  },
]
`;

exports[`should work 7`] = `
[
  [
    {
      "name": "foo",
    },
  ],
  [
    {
      "name": "bar",
    },
  ],
]
`;

exports[`should work 8`] = `
[
  {
    "type": "return",
    "value": Promise {},
  },
  {
    "type": "return",
    "value": Promise {},
  },
]
`;
